### PR TITLE
Added endpoints for animal and service

### PR DESCRIPTION
**Pull request by tanveersaleem786**
Created endpoints for animal and service that can be used by Front End Team.

### CODE REVIEW

**Review by wilcoxva — COMMENTED** [review]
looking good!

File: api/animal/animalModel.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+const findById = async (id) => {
+  return db('animals').where({ id }).first().select('*');
+};
+
+const create = async (animal) => {
+  return db('animals').insert(animal).returning('*');
+};
+
+const update = (id, animal) => {
+  return db('animals').where({ id: id }).first().update(animal).returning('*');
```
Comment (TylerAlsop) [inline api/animal/animalModel.js:20]: I like how you're also returning everything for that animal id in both the create and update.

File: api/animal/animalRouter.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  isAnimalUnique(),
+  async (req, res, next) => {
+    try {
+      const payload = {
+        animal_type: req.body.animal_type,
+      };
+      const animal = await animalModel.create(payload);
+      return res
+        .status(200)
+        .json({ message: 'Animal created', animal: animal[0] });
```
Comment (TylerAlsop) [inline api/animal/animalRouter.js:161]: Good clear message upon success.

File: api/animal/animalRouter.js [hunk 3]
```diff
@@ -0,0 +1,273 @@
+const express = require('express');
+const authRequired = require('../middleware/authRequired');
+const animalModel = require('./animalModel');
+const router = express.Router();
+const {
+  validateAnimalID,
+  validateAnimalData,
+  isAnimalUnique,
```
Comment (TylerAlsop) [inline api/animal/animalRouter.js:8]: Great job adding all the validations.

File: api/animal/animalRouter.js [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+  '/:id',
+  authRequired,
+  validateAnimalID(),
+  async (req, res, next) => {
+    try {
+      const id = req.params.id;
+      await animalModel.remove(id);
+      return res
+        .status(200)
+        .json({ message: `Animal '${id}' was deleted.`, animal: req.animal });
```
Comment (TylerAlsop) [inline api/animal/animalRouter.js:266]: I like that you added the animal id to the message upon a successful delete.

File: api/app.js [hunk 5]
```diff
@@ -55,6 +57,8 @@ app.use(cookieParser());
 // application routes
 app.use('/', indexRouter);
 app.use(['/profile', '/profiles'], profileRouter);
+app.use(['/animal', '/animals'], animalRouter);
+app.use(['/service', '/services'], serviceRouter);
```
Comment (TylerAlsop) [inline api/app.js:61]: I like that there are multiple options for these url endpoints. Especially right now while we're deep into production right now.

File: api/middleware/animal.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
+const Animals = require('../animal/animalModel');
+const db = require('../../data/db-config');
+
+module.exports = {
+  validateAnimalID: () => {
+    return async (req, res, next) => {
+      try {
+        const id = req.body.id || req.params.id;
```
Comment (TylerAlsop) [inline api/middleware/animal.js:8]: Smart adding a default here.

File: api/middleware/animal.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  validateAnimalData: () => {
+    return async (req, res, next) => {
+      if (
+        req.body.constructor === Object &&
+        Object.keys(req.body).length === 0
+      ) {
+        res.status(404).json({ message: 'Missing animal data' });
+      } else if (!req.body.animal_type) {
+        res.status(404).json({ message: 'Missing animal type field' });
```
Comment (TylerAlsop) [inline api/middleware/animal.js:30]: Good response failure messages.

File: api/middleware/animal.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        const animal_qry = db('animals').where(
+          'animal_type',
+          req.body.animal_type
+        );
+        if (animal_id) {
+          animal_qry.whereNot('id', animal_id); // in case of update
+        }
+
+        if (await animal_qry.first()) {
+          res.status(404).json({ message: 'Animal alreday exist' });
```
Comment (TylerAlsop) [inline api/middleware/animal.js:50]: Minor spelling errors here. Should read, 'Animal already exists'
We can probably change this ourselves. Possibly right after merging?

File: api/middleware/service.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    return async (req, res, next) => {
+      try {
+        const service_id = req.body.id;
+        const service_qry = db('services').where('name', req.body.name);
+        if (service_id) {
+          service_qry.whereNot('id', service_id); // in case of update
+        }
+
+        if (await service_qry.first()) {
+          res.status(404).json({ message: 'Service alreday exist' });
```
Comment (TylerAlsop) [inline api/middleware/service.js:47]: Again some minor spelling errors here. Should read, 'Service already exists'
We can probably change this ourselves. Possibly right after merging?

File: api/service/serviceModel.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+const findById = async (id) => {
+  return db('services').where({ id }).first().select('*');
+};
+
+const create = async (animal) => {
+  return db('services').insert(animal).returning('*');
+};
+
+const update = (id, animal) => {
+  return db('services').where({ id: id }).first().update(animal).returning('*');
```
Comment (TylerAlsop) [inline api/service/serviceModel.js:20]: Again I like how you are returning everything for services in create and update, just like you did for animals.

File: api/service/serviceRouter.js [hunk 11]
```diff
@@ -0,0 +1,279 @@
+const express = require('express');
+const authRequired = require('../middleware/authRequired');
+const serviceModel = require('./serviceModel');
+const router = express.Router();
+const {
+  validateServiceID,
+  validateServiceData,
+  isServiceUnique,
```
Comment (TylerAlsop) [inline api/service/serviceRouter.js:8]: Good job with the validating like you did with animals.

**Review by TylerAlsop — APPROVED** [review]
Looks good! It was very cleanly done and DRY. Only saw a couple of spelling errors that can be easily fixed.

**Pull request merged**